src/spac/transformations.py

```diff
@@ -101,6 +102,92 @@ def phenograph_clustering(
     adata.uns["phenograph_features"] = features
 
 
+def knn_clustering(
+        adata,
+        features,
+        label,
+        layer=None,
+        k=50,
+        output_annotation="knn",
+        associated_table=None,
+        **kwargs):
+    """
+    Calculate knn clusters using sklearn KNeighborsClassifier
+
+    The function will add these two attributes to `adata`:
+    `.obs["knn"]`
+        The assigned int64 class labels by KNeighborsClassifier
+
+    `.uns["knn_features"]`
+        The features used to calculate the knn clusters
+
+    Parameters
+    ----------
+    adata : anndata.AnnData
+       The AnnData object.
+
+    features : list of str
+        The variables that would be included in creating the phenograph
+        clusters.
+    
+    label : str [andata.obs.label]
+        The name of the label used for classifying the data
+
+    layer : str, optional
+        The layer to be used in calculating the phengraph clusters.
+
+    k : int, optional
+        The number of nearest neighbor to be used in creating the graph.
+
+    output_annotation : str, optional
+        The name of the output layer where the clusters are stored.
+
+    associated_table : str, optional
+        If set, use the corresponding key `adata.obsm` to calcuate the
+        clustering. Takes priority over the layer argument.
+
+    Returns
+    -------
+    adata : anndata.AnnData
+        Updated AnnData object with the knn clusters
+        stored in `adata.obs[output_annotation]`
+    """
+
+    # 1 read in data, validate labels in the call here
+    _validate_transformation_inputs(
+        adata=adata,
+        layer=layer,
+        associated_table=associated_table,
+        features=features,
+        label=label,
+    )
+
+    if not isinstance(k, int) or k <= 0:
+        raise ValueError("`k` must be a positive integer")
+    
+    data = _select_input_features(
+        adata=adata,
+        layer=layer,
+        associated_table=associated_table,
+        features=features
+    )
+
+    # 2 we must split the labeled data from the unlabeled data
+    y = adata.obs[label]
+    y_mask = y != "no_label"

```

Comment:
let's add a check to make sure that y_mask is non-trivial (neither all the data is labeled nor none of it). if the annotation is faulty (no labeled samples or all labeled, throw an error)